api/src/main/java/org/openmrs/api/impl/OrderServiceImpl.java

```diff
@@ -325,28 +326,38 @@ private void discontinueExistingOrdersIfNecessary(Order order) {
 		    null);
 		boolean isDrugOrderAndHasADrug = DrugOrder.class.isAssignableFrom(getActualType(order))
 		        && ((DrugOrder) order).getDrug() != null;
+		Order orderToBeDiscontinued = null;
 		for (Order activeOrder : orders) {
 			if (!getActualType(order).equals(getActualType(activeOrder))) {
 				continue;
 			}
-			boolean shouldMarkAsDiscontinued = false;
 			//For drug orders, the drug must match if the order has a drug
 			if (isDrugOrderAndHasADrug) {
-				DrugOrder drugOrder1 = (DrugOrder) order;
-				DrugOrder drugOrder2 = (DrugOrder) activeOrder;
-				if (OpenmrsUtil.nullSafeEquals(drugOrder1.getDrug(), drugOrder2.getDrug())) {
-					shouldMarkAsDiscontinued = true;
-				}
+				if(orderToBeDiscontinued==null){
+					orderToBeDiscontinued = checkDrugOrdersForDiscontinuing((DrugOrder) order, (DrugOrder) activeOrder);
+				}else{
```

Comment:
This logic is actually not exactly worng because checkDrugOrdersForDiscontinuing could return null, the else clause on line 338 above assumes checkDrugOrdersForDiscontinuing won't return null, i will fix it